### ISSUE TITLE
[twice-regressed] Hebrew chars not displayed at all

### ISSUE THREAD

**Issue by ousia** [ruffle-rs/ruffle#8885]
### Describe the bug

This comes from #5839.

https://github.com/ruffle-rs/ruffle/files/6273851/clock.zip should have chars displayed as:

![clock-fs8](https://user-images.githubusercontent.com/203767/209017646-45fc9342-0c50-44c1-86eb-4aa165df2e32.png)

#5839 was solved by #5900.

But using current nightly from 2022-12-21, Hebrew characters aren’t displayed at all:
 
![](https://user-images.githubusercontent.com/18469146/209016091-1f90bfe7-7877-4786-9b3a-f2d8ce1a41e1.png)

### Expected behavior

Hebrew characters should be displayed (ideally, R2L and not L2R [as already reported in #3921]).

Many thanks for your help.

### Affected platform

Online demo

### Operating system

Fedora 35

### Browser

Firefox 107

**Comment by ousia** [comment]
Screenshot from browser (since I haven’t generated the Windows image myself):

![clse-fs8](https://user-images.githubusercontent.com/203767/209019124-20d258d0-eab2-4a5f-af4a-6116435c9b61.png)

**Comment by danielhjacobs** [comment]
You can try to find when it regressed using https://ruffle-regressions.github.io/.

If you don't, someone else probably will, though I'm busy right now.

**Comment by ousia** [comment]
@danielhjacobs,

many thanks for the info on https://ruffle-regressions.github.io (I didn’t know there was such a site).
According to the images:

![displayed-l2r-fs8](https://user-images.githubusercontent.com/203767/209083690-648c9f94-76ec-409e-8b6d-a72bb2e11833.png)

Nightly from 2022-09-08 seems to be the culprit:

![notatall-displayed-fs8](https://user-images.githubusercontent.com/203767/209083696-25a94576-975e-4a01-88a5-56bea849b4ad.png)

So the code is https://github.com/ruffle-rs/ruffle/compare/nightly-2022-09-07...nightly-2022-09-08.

The only PR related to text seems to be #7853 (but I may be wrong here, since there seems to be no EditText in the SWF).

@dhtdht020, could you also confirm whether texts are displayed R2L? On Linux, texts seem to be the same.

![two-texts-fs8](https://user-images.githubusercontent.com/203767/209086914-29bf3042-26ee-4f8b-a0eb-8ce25c754f9f.png)

Many thanks for your help.

**Comment by n0samu** [comment]
Looks like the characters are displayed incorrectly in the Linux Flash projector but correctly in the Windows Flash projector. Here's how the bottom-left text looks for me in the Windows projector:
![image](https://user-images.githubusercontent.com/71368227/209092606-a8c1b93c-7100-470f-86c4-dd6121c67b5d.png)

**Comment by n0samu** [comment]
> The only PR related to text seems to be #7853 (but I may be wrong here, since there seems to be no EditText in the SWF).

That is indeed correct - I've bisected the problem to commit 84481c28040537c76fe7159d601cec5b87fd7672 of that PR. The SWF actually creates all of its text fields at runtime using `createTextField`.

**Comment by ousia** [comment]
@n0samu,

I get the same output with the Windows Flash projector using Wine in Linux.

![wine-clock-fs8](https://user-images.githubusercontent.com/203767/209112753-5a577334-e09a-4238-9dd5-71300ce9e0c8.png)

Many thanks for your help.

**Comment by ousia** [comment]
My knowledge or other scripts (and languages) is limited to L2R.

So I didn’t know that Wine had problems with R2L.

**Comment by dhtdht020** [comment]
I can confirm this is still an issue as of Nightly 2023-09-08.
![image](https://github.com/ruffle-rs/ruffle/assets/18469146/6fd43c7a-c027-4fd0-ab24-19e87d7dc94b)
Could also replicate this exact issue on Linux through online demo on firefox, so does not seem platform-specific.

> @dhtdht020, could you also confirm whether texts are displayed R2L? On Linux, texts seem to be the same.

@ousia Sorry for not responding, GitHub did not notify me about the mention. The text in the Flash Player running under wine is displayed in the wrong order similarly to the problem presented in #3921, unlike Flash Player running on a proper Windows installation which displays the text correctly.

**Comment by n0samu** [comment]
After looking more into this, it unfortunately seems to be expected behavior, not a bug. The SWF contains an Arial font with all of the needed characters, but does not actually use the embedded font. It only works in Flash Player because it is able to load the Arial font installed on the user's device. Now that #14535, it works in the desktop app, at least. To work on the web, either the code would need to be modified, or a second SWF containing the required font would need to be created and Ruffle would need to be configured to use it. [See the FAQ for instructions](https://github.com/ruffle-rs/ruffle/wiki/Frequently-Asked-Questions-For-Users#how-do-i-change-the-fonts-used-by-flash-on-my-website).

**Comment by n0samu** [comment]
Yes, currently the web environment does not support loading fonts from the user's device.

**Comment by danielhjacobs** [comment]
> is there no way for the web to load device fonts, such as the desktop app does?

On self-hosted web, you can provide your own fonts and configure ruffle to treat these as if they were device fonts: https://github.com/ruffle-rs/ruffle/wiki/Frequently-Asked-Questions-For-Users#how-do-i-change-the-fonts-used-by-flash-on-my-website.

There is no widely supported web API to allow Ruffle to access the glyphs of your system's installed fonts, so we can't do this the same way it was done with Flash. The closest API is https://developer.mozilla.org/en-US/docs/Web/API/Local_Font_Access_API, but it's Chromium only and has some issues with the design which make it non-ideal to use.
